Spike/plugin macro api

keathley opened the pull request:
Working out the details for the new regex based plugin api.

I'm doing most of the work with the `respond` and `hear` macros in the Butler.Plugin module.

I'd still like to revisit the way that each plugin is called from the bot. We've talked about moving away from GenEvent, which I still think might be a possibility. I'm also not sure that we really need the `Bot` module to function as a process. I'm really beginning to think that its just a bottleneck. I do like the api that it provides but I'm not sure that we couldn't just call the plugins directly. Especially since there really isn't state that the bot needs to keep track of. I could be wrong about that as well.

There are definitely going to be issues with the Task solution. But I just thought that I would give it a try. If you think it makes more sense to use the GenEvent api then just let me know and I can take a look at it.

Also yay for tests.

keathley committed:
WIP

keathley committed:
Refactor respond to use a macro.

Users can use regex to match against a message.

Respond adds the bots name to the regex and returns any capture groups
to the plugin.

keathley committed:
Update call sites to use Tasks to call all the plugins

bbrock25 commented on lib/butler/plugin.ex:
```diff
@@ -3,77 +3,117 @@ defmodule Butler.Plugin do
     end
   end
+
+  defp source([{:<<>>, _, [string]}, []]) do
```
what is going on here?

```
:<<>>
```

keathley replied:
thats the quoted version of a regex pattern

keathley committed:
Refactor plugin architecture

Why:

Plugins should run for indeterminant amounts of time and also be
non-blocking.

This change addresses the need by:

Plugins now get their own task to run in. Plugins have also been
refactored to take a connection object (really a `Butler.Message`). This
struct contains everything that the plugin needs in order to respond to
the proper user and channel. The plugin then responds with a
`Butler.Response` built from the text response and from the data passed
in the initial connection struct.

keathley committed:
Remove dead code

keathley committed:
Use NullAdapter in test mode

keathley committed:
Remove plugin until we can update to new syntax

keathley commented:
@bbrock25, @colinrymer: I've updated it so each plugin can now run in its own task. These tasks don't have a timeout since they aren't applying backpressure. This allows plugins to respond multiple times during its lifespan. It also removes the bottleneck of waiting for each plugin to finish before moving on.

Lemme know what you think. If you think it looks good then I think the last thing to do is just to get the slack adapter updated to use the new interface and we can call this good.

bbrock25 commented on lib/butler/adapters/console.ex:
```diff
@@ -37,10 +36,14 @@ defmodule Butler.Adapters.Console do
+  def format_response(_), do: ""
+
+  defp mention_user(user) do
+    "@#{user}: "
```
Is this to mimic the slack api?

keathley replied:
We're just simplifying the api so we don't have to lift the value out of the tuple.

bbrock25 commented on lib/butler/bot.ex:
```diff
@@ -1,39 +1,21 @@
-  def start_link(event_manager) do
-    GenServer.start_link(__MODULE__, {event_manager}, [name: __MODULE__])
+  def notify(msg) do
+    @plugins
```
I like this :)

bbrock25 commented on lib/butler/plugins/cowsay.ex:
```diff
@@ -0,0 +1,9 @@
+defmodule Butler.Plugins.Cowsay do
+  use Butler.Plugin
```
can we publish the butler plugin api to hex so that external plugins can just require it?

keathley replied:
As far as I know the plugins should be able to reference the main butler repo and then pull in just the `Butler.Plugin` module. If thats not the case we should try to figure out why.

bbrock25 replied:
Ok, very cool. I'll update the lil cowsay plugin and see what I learn.

keathley committed:
Get slack adapter working...ish

keathley committed:
Update both slack and console adapters

bbrock25 commented:
Hurrah! Looking forward to playing with this
